File: Day-49/FlattenMDLL.java

public class FlattenMDLL {

class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
};



    public Node flatten(Node head) {
        Node temp=head;
        while(temp!=null)
        {
            if(temp.child!=null)
            {
                Node after=temp.next;
                temp.next=flatten(temp.child);
                temp.next.prev=temp;
                temp.child=null;
                while(temp.next!=null)
                {
                    temp=temp.next;
                }
                if(after!=null)
                {
                    temp.next=after;
                    after.prev=temp;
                }
            }
            temp=temp.next;
        }
        return head;
    }

}

File: Day-49/FlattenMDLL.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Optional[Node]') -> 'Optional[Node]':
        curr=head
        while curr:
            if curr.child:
                after=curr.next
                curr.next=self.flatten(curr.child)
                curr.next.prev=curr
                curr.child=None
                while curr.next:
                    curr=curr.next
                if(after):
                    curr.next=after
                    after.prev=curr
            curr=curr.next
        return head
        